File: es10/es10.py
#!/usr/bin/env python3

import sys
import re

class Light(object):
    def __init__(self, x, y, vx, vy):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy

    def step_forward(self):
        self.x += self.vx
        self.y += self.vy

    def step_backwards(self):
        self.x -= self.vx
        self.y -= self.vy

class Extents(object):
    def __init__(self):
        self.xmin = 1000000000
        self.ymin = 1000000000
        self.xmax = -1000000000
        self.ymax = -1000000000

    def width(self):
        return self.xmax - self.xmin + 1

    def height(self):
        return self.ymax - self.ymin + 1

def find_area(lights):
    e = Extents()
    for l in lights:
        if e.xmin > l.x:
            e.xmin = l.x
        if e.xmax < l.x:
            e.xmax = l.x
        if e.ymin > l.y:
            e.ymin = l.y
        if e.ymax < l.y:
            e.ymax = l.y
    return e, e.width() * e.height()

pattern = re.compile(r"position=<\s*([-]?\d+),\s*([-]?\d+)> velocity=<\s*([-]?\d+),\s*([-]?\d+)>")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    lights = []
    try:
        with open(sys.argv[1], "r") as f:
            for line in f:
                m = pattern.match(line)
                if m:
                    params = [int(m.group(x+1)) for x in range(4)]
                    lights.append(Light(*params))
                else:
                    print(line)
    except:
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    print("Read {} light points".format(len(lights)))

    min_area = 100000000000000000
    min_ext = None
    min_sec = 0
    cur_sec = 0
    while True:
        ext, area = find_area(lights)
        if min_area > area:
            min_area = area
            min_ext = ext
            min_sec = cur_sec
        elif min_area < area:
            break

        for l in lights:
            l.step_forward()

        cur_sec += 1

    print("Minimum reached after {} seconds, area {}".format(min_sec, min_area))
    print("Map size {}x{} starting at ({},{})".format(
        min_ext.width(), min_ext.height(), min_ext.xmin, min_ext.ymin))

    while cur_sec > min_sec:
        cur_sec -= 1
        for l in lights:
            l.step_backwards()

    for y in range(min_ext.ymin, min_ext.ymax+1):
        line = []
        for x in range(min_ext.xmin, min_ext.xmax+1):
            found = False
            for l in lights:
                if l.x == x and l.y == y:
                    found = True
                    line.append("#")
                    break
            if not found:
                line.append(" ")
        print("".join(line))

File: es10/es10.go
package main

import (
	"fmt"
	"os"
)

const MaxInt64 = int64(^uint64(0) >> 1)
const MinInt64 = -MaxInt64 - 1

type Light struct {
	X, Y   int64
	Vx, Vy int64
}

func (l *Light) stepForward() {
	l.X += l.Vx
	l.Y += l.Vy
}

func (l *Light) stepBackwards() {
	l.X -= l.Vx
	l.Y -= l.Vy
}

type Extents struct {
	Xmin, Ymin int64
	Xmax, Ymax int64
}

func findArea(lights []*Light) (Extents, int64) {
	e := Extents{MaxInt64, MaxInt64, MinInt64, MinInt64}
	for _, l := range lights {
		if e.Xmin > l.X {
			e.Xmin = l.X
		}
		if e.Xmax < l.X {
			e.Xmax = l.X
		}
		if e.Ymin > l.Y {
			e.Ymin = l.Y
		}
		if e.Ymax < l.Y {
			e.Ymax = l.Y
		}
	}
	return e, (e.Xmax - e.Xmin + 1) * (e.Ymax - e.Ymin + 1)
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	lights := []*Light{}
	for {
		var x, y, vx, vy int64
		_, err = fmt.Fscanf(input, "position=<%d,%d> velocity=<%d,%d>\n",
			&x, &y, &vx, &vy)
		if err != nil {
			break
		}
		lights = append(lights, &Light{x, y, vx, vy})
	}

	fmt.Printf("Read %d light points\n", len(lights))

	minArea := MaxInt64
	minSec := 0
	minExt := Extents{}
	curSec := 0
	for ; ; curSec++ {
		ext, area := findArea(lights)
		if minArea > area {
			minArea = area
			minSec = curSec
			minExt = ext
		} else if minArea < area {
			break
		}
		for _, l := range lights {
			l.stepForward()
		}
	}

	fmt.Printf("Minimum reached after %d seconds, area %d\n", minSec, minArea)

	for ; curSec > minSec; curSec-- {
		for _, l := range lights {
			l.stepBackwards()
		}
	}

	fmt.Printf("Map size %dx%d starting at(%d,%d)\n",
		minExt.Xmax-minExt.Xmin+1,
		minExt.Ymax-minExt.Ymin+1,
		minExt.Xmin, minExt.Ymin)

	for y := minExt.Ymin; y <= minExt.Ymax; y++ {
		for x := minExt.Xmin; x <= minExt.Xmax; x++ {
			found := false
			for _, l := range lights {
				if l.X == x && l.Y == y {
					fmt.Printf("#")
					found = true
					break
				}
			}
			if !found {
				fmt.Printf(" ")
			}
		}
		fmt.Println()
	}

	return 0
}
